### PR TITLE
Apply carriedVelocity unless substantially different

### LINKED ISSUE #77043: ListView ScrollPhysics different from native scroll physics when pausing in between scrolls
## Steps to Reproduce

1. Pause in between scrolls on a ListView.
2. Observe subsequent scrolls adding accelerations from previous, eventually looking like "jumps".

**Expected results:**

Subsequent scrolls should not add accelerations from previous scrolls.
It should start from 0, like the initial scroll behavior.

Clubhouse example:

https://user-images.githubusercontent.com/3824737/109612749-8ce1dd00-7b73-11eb-994d-16f153f20772.mp4

**Actual results:** 

Subsequent scrolls add accelerations from previous scrolls, causing unpleasant jumps.

My flutter app example:

https://user-images.githubusercontent.com/3824737/109613052-fc57cc80-7b73-11eb-92aa-69b7012dd25b.mp4

### LINKED ISSUE #79277: Carried velocity is incorrectly applied on iOS
On iOS with `BouncingScrollPhysics` the carried velocity that is applied at the end of a scroll activity seems to be incorrectly applied. It will always apply the velocity even if the last fling's velocity was substantially less than previous flings. This will cause the visible effect of the scrollview _accelerating_ after a slower fling. 

## Steps to Reproduce

These steps assumes the usage of the [scroll_overlay](https://github.com/flutter/platform_tests/tree/master/scroll_overlay) platform test

For the unxperienced finger this is a bit challenging to _intentionally_ trigger.  Therefor I have included two ways to trigger it and a video at the end (as it was difficult to show how to reproduce with just a screen-capture). 

1. Build momentum by flinging 3-4 times a bit fast
2. Then while continuing to fling slow down the velocity of the flings
3. Make one final fling that is slowing that the flings you made 3-4 times
4. Observe that the scrollview's velocity is a lot more than the native scroll view and in some cases will even have a larger velocity than what you finger had while in contact with the screen.

https://user-images.githubusercontent.com/1550685/112858768-79faf380-90b2-11eb-8133-3d07029d486d.mp4

**Expected results:** 

Somewhat similar velocity to the native iOS scrollview when lifting the finger for the last time

**Actual results:**

The scrollview simply added the previously carried momentum to the velocity of the current fling. Causing the effect of making the scrollview speed up and scroll for much longer than the native view. This is especially jarring when you barely have any velocity while lifting the finger with carried momentum, like trying to stop a scroll by swiping just a little bit. 

```
[✓] Flutter (Channel beta, 1.25.0-8.3.pre, on macOS 11.1 20C69 darwin-x64, locale en-GB)
    • Flutter version 1.25.0-8.3.pre at /Users/trymnilsen/development/flutter
    • Framework revision 5d36f2e7f5 (2 months ago), 2021-01-14 15:57:49 -0800
    • Engine revision 7a8f8ca02c
    • Dart version 2.12.0 (build 2.12.0-133.7.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/trymnilsen/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio 4.1 Preview.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio 4.1 Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Android Studio
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (1 available)
    • iPhone (mobile) • <id redacted> • ios • iOS 14.4.2

• No issues found!
```

### PR DESCRIPTION
**Pull request by trymnilsen**
On iOS the `carriedVelocity` (if it should be retained) is always added when a drag is ended. This can cause the unexpected behaviour of the scrollview speeding up after a fling, something that is not the expected behaviour in a native iOS app. This PR seeks to fix this issue by adding a check that the incoming velocity is larger than a factor of the carried velocity before applying it.

Attached two videos show the before and after behaviour

https://user-images.githubusercontent.com/1550685/113015889-3f5e8d00-917e-11eb-92be-6f8d2278c714.mp4

https://user-images.githubusercontent.com/1550685/113015759-2655dc00-917e-11eb-8a7a-0e3df19b28e6.mp4

Related Issues
Fixes #77043
Fixes #79277

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

I intend to add a test for this as well, but I wanted to get some feedback on the suggested solution first, seeing as this is my first contribution to flutter.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 1]
```diff
@@ -274,6 +274,11 @@ class ScrollDragController implements Drag {
   static const Duration momentumRetainStationaryDurationThreshold =
       Duration(milliseconds: 20);
 
+  /// The factor multiplied with the carriedVelocity to check if it should be
+  /// added to added to the velocity of an ended scroll gesture.
```
Comment (gaaclarke) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:278]: s/added to added/added to

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 2]
```diff
@@ -274,6 +274,11 @@ class ScrollDragController implements Drag {
...
+  /// Decided by fair eyeballing with the scroll_overlay platform test
```
Comment (gaaclarke) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:279]: Use a complete sentence please.

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 3]
```diff
@@ -391,8 +396,17 @@ class ScrollDragController implements Drag {
...
     // Build momentum only if dragging in the same direction.
-    if (_retainMomentum && velocity.sign == carriedVelocity!.sign)
+    // Check that the incomming velocity is more than
+    // a [momentumRetainVelocityThresholdFactor]'th of the carriedVelocity to
+    // avoid adding the carriedVelocity to the current velocity if the finger
+    // left the screen with a substantially smaller speed giving the effect of
+    // the scrollview carrying on in a much larger speed than the finger left with
+    if (_retainMomentum &&
+        velocity.sign == carriedVelocity!.sign &&
+        (carriedVelocity! == 0 || velocity.abs() > carriedVelocity!.abs() * momentumRetainVelocityThresholdFactor)
```
Comment (gaaclarke) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:406]: There's no need to add the carriedVelocity if it is zero.

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 4]
```diff
@@ -391,8 +396,17 @@ class ScrollDragController implements Drag {
     _lastDetails = details;
 
     // Build momentum only if dragging in the same direction.
-    if (_retainMomentum && velocity.sign == carriedVelocity!.sign)
+    // Check that the incomming velocity is more than
+    // a [momentumRetainVelocityThresholdFactor]'th of the carriedVelocity to
+    // avoid adding the carriedVelocity to the current velocity if the finger
+    // left the screen with a substantially smaller speed giving the effect of
+    // the scrollview carrying on in a much larger speed than the finger left with
```
Comment (gaaclarke) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:403]: Complete sentences please.

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 3 above)
Comment (gaaclarke) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:406]: I think this sounds good, @xster might have an opinion since he wrote this orignally.

**Review by gaaclarke — COMMENTED** [review]
Thanks for the contribution.  I think the concept sounds fine.  Can you please add a test and clean up the comments so they are complete sentences?

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 5]
```diff
@@ -274,6 +274,11 @@ class ScrollDragController implements Drag {
   static const Duration momentumRetainStationaryDurationThreshold =
       Duration(milliseconds: 20);
 
+  /// The factor multiplied with the carriedVelocity to check if it should be
```
Comment (xster) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:277]: I didn't understand this sentence :D 

Feel free to use longer/multiple sentences to describe it. Also, an example can help illustrate if needed.

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:406]: this if clause is too long. Break this line or part of this line out into a local variable before the if. It also has the added advantage that that variable name is then self documenting.

**Review by xster — COMMENTED** [review]
Sensible addition. Thanks for making the videos to help illustrate the point. This needs a test added somewhere around https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/scrollable_test.dart#L160.

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 2 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: Rephrased the whole comment

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 3 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: Added each condition as a local

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 4 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: Rewritten this

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 3 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: 👍

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 5 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: Rephrased this and added an example

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 3 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: Created locals, let me know if you have any comments on their naming

File: packages/flutter/lib/src/widgets/scroll_activity.dart (hunk 1 above)
Comment (trymnilsen) [reply packages/flutter/lib/src/widgets/scroll_activity.dart]: rephrased this

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 6]
```diff
@@ -274,6 +274,12 @@ class ScrollDragController implements Drag {
   static const Duration momentumRetainStationaryDurationThreshold =
       Duration(milliseconds: 20);
 
+  /// The minimum amount of velocity needed to apply the [carriedVelocity] at
+  /// the end of a drag. Expressed as a factor. For example with a [carriedVelocity]
+  /// of 2000, we will need a velocity of at least 1000 to apply the [carriedVelocity]
+  /// as well. Decided by fair eyeballing with the scroll_overlay platform test.
```
Comment (xster) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:280]: Describe what happens if it doesn't meet the threshold. Looking at the video, it seems like a new fling velocity is re-evaluated based on the speed of the last swipe?

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 7]
```diff
@@ -390,9 +396,15 @@ class ScrollDragController implements Drag {
...
     _lastDetails = details;
 
-    // Build momentum only if dragging in the same direction.
-    if (_retainMomentum && velocity.sign == carriedVelocity!.sign)
-      velocity += carriedVelocity!;
+    if (_retainMomentum) {
+      // Build momentum only if dragging in the same direction.
+      final bool isFlingingInSameDirection = velocity.sign == carriedVelocity!.sign;
+      // Build momentum only if the velocity of the last drag was not substantially lower than the carried momentum.
+      final bool isVelocityNotSubstantiallyLessThanCarriedMomentum = velocity.abs() > carriedVelocity!.abs() * momentumRetainVelocityThresholdFactor;
```
Comment (xster) [inline packages/flutter/lib/src/widgets/scroll_activity.dart:403]: We generally keep lines at 80 char. https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters 

Break into 2 lines.

File: packages/flutter/test/widgets/scrollable_test.dart [hunk 8]
```diff
@@ -170,6 +170,24 @@ void main() {
...
+    await pumpTest(tester, debugDefaultTargetPlatformOverride);
+    await tester.fling(find.byType(Scrollable), const Offset(0.0, -dragOffset), 1000.0);
+    await tester.pump(); // trigger fling
+    await tester.pump(const Duration(milliseconds: 10));
+    // Repeat the exact same motion to build momentum.
+    await tester.fling(find.byType(Scrollable), const Offset(0.0, -dragOffset), 1000.0);
+    await tester.pump(); // trigger the second fling
+    await tester.pump(const Duration(milliseconds: 10));
+    // Make a final fling that is much slower.
+    await tester.fling(find.byType(Scrollable), const Offset(0.0, -dragOffset), finalFlingSpeed);
```
Comment (xster) [inline packages/flutter/test/widgets/scrollable_test.dart:184]: Be consistent. Either have fast and slow fling speeds in local variables or just write the numbers in

**Review by xster — APPROVED** [review]
LGTM! Thanks for your contribution

**Comment by trymnilsen** [comment]
Added the final changes you suggested. Let me know if you have any further comments. Thanks for the review.

**Comment by xster** [comment]
Awesome work, thanks!!

**Pull request merged**